### DISCUSSION TITLE
Product Summary Block: character limit and rendering behavior of different scenarios

### DISCUSSION THREAD

**Discussion by gigitux** [woocommerce/woocommerce#44260] [WooCommerce, Blocks, and Site Editing]
## Introduction
The Product Summary block, a variant of the Post Excerpt block, possesses a character limit that may not be immediately apparent to users. This lack of highlighting leads to confusion among users, who subsequently report issues of text being truncated on the front end.

### Scenario 1 - no short description

When a product lacks a short description, on the blockified template the Product Summary truncates the product description to a maximum of 55 words. This can be observed in the following example:

#### Product Setting

![image](https://github.com/woocommerce/woocommerce/assets/4463174/2b28d167-9aee-4e02-bd05-ce5d693e3b1b)

#### Frontend

| Blockified Template | Classic Template |
|--------|--------|
|![image](https://github.com/woocommerce/woocommerce/assets/4463174/9a199f7a-b547-4376-b8ad-d85a17bbb69e) | ![image](https://github.com/woocommerce/woocommerce/assets/4463174/ce17d9f6-cbe9-4951-981b-1cd485fb693f)| 

You can notice that on the blockified template, the description is present twice because the Post Excerpt renders the first 55 words of the Post. ([docs](https://wordpress.org/documentation/article/page-post-settings-sidebar/#excerpt)).

#### Questions
- Is it acceptable to render the same description twice? While not an SEO expert, duplicating content on the same page may impact SEO performance.
- If yes, should the truncated version be displayed?

### Scenario 2 - with short description

On the other hand, when a product includes a short description, the rendering is as follows:

#### Product Setting

![image](https://github.com/woocommerce/woocommerce/assets/4463174/0403c84e-b899-484a-a8ac-68d00a07ea8f)

#### Frontend

| Blockified Template | Classic Template |
|--------|--------|
| ![image](https://github.com/woocommerce/woocommerce/assets/4463174/13c6f822-a8a7-4c7a-91bc-41fa41083361) | ![image](https://github.com/woocommerce/woocommerce/assets/4463174/a43e7496-0bc0-4464-9a43-cfb262a43776)| 

In this scenario, the Classic Template displays the entire "short description," while the Product Summary block renders only the initial 55 words.

#### Question

- Should the Product Summary truncate the short description or display the entire description, similar to the Classic Template?

### Other potential improvments & technical limitations

I want to highlight that the Product Summary is a variation of the Post Excerpt, so we don't have complete control over the block's behavior. If we want to improve the block, we have to contribute upstream; otherwise, we should create a dedicated block.

When considering improvements, @jarekmorawski presented a mockup that aligns well with both the Product Summary variation and the Post Excerpt (except for the "post" word). The mockup:

![image](https://github.com/woocommerce/woocommerce/assets/4463174/7d771a24-eb36-40a0-b422-24f4fafb62e9)

This design appears suitable for both the Product Summary variation and the general Post Excerpt use case.

Also, @jarekmorawski suggested improving the copy to be more specific for product context. Unfortunately, if we stick with the variation, we can't make it more specific. This is the PR that I did to improve the copy on the editor side and make it more general: https://github.com/WordPress/gutenberg/pull/50167

### Short-Term Solution

As a short-term solution, I would suggest increasing the length passing a custom `excerptLenght` attributes ([source code](https://github.com/WordPress/gutenberg/blob/23bdd6679d73bca1f73c614644057c0f4e0c133f/packages/block-library/src/post-excerpt/block.json/#L20-L23)) when we register the Product Summary variation. ([source code](https://github.com/woocommerce/woocommerce/blob/78e38abef44f94c7a93bcafb47589bea82e5c34a/plugins/woocommerce-blocks/assets/js/blocks/product-query/variations/elements/product-summary.tsx/#L19-L26)).

### Long-Term Solution

As a long-term solution, we should analyze the current status of the block, consider all areas for improvement, and ensure that the Post Excerpt has the same requirements. If yes, we can contribute upstream; otherwise, it could make sense to create a dedicated block.

What do you think?

**Comment by Aljullu** [comment]
Thanks for writing this up, @gigitux! I agree with your proposed solutions, so I don't have much to add.

> If yes, we can contribute upstream; otherwise, it could make sense to create a dedicated block.

From what I understand, the only change we would need to upstream is adding the option to "Show post content if empty" from Jarek's designs, is that correct? Because scenario 2 can be fixed with the `excerptLength` attribute without requiring any upstream changes, right?

In that case, I think we should create an issue in Gutenberg to see if adding a "Show post content if empty" attribute is something they would be interested in. IMO, it would be a nice feature to have in Gutenberg. :slightly_smiling_face:  We could then contribute with a PR.

**Reply by gigitux** [reply]
> Because scenario 2 can be fixed with the excerptLength attribute without requiring any upstream changes, right?

Not really, because with the excerptLength attribute, we put a constraint (even if the constraint is very high). We should implement upstream a way to allow users/developers to disable the truncation: does it make sense?

**Reply by Aljullu** [reply]
Oh, you're right. Setting a high number in `excerptLength` is more a work-around than a real fix.

**Comment by kmanijak** [comment]
Mentioning two more issues related to this problem raised by community:
- https://github.com/woocommerce/woocommerce/issues/44185
- https://github.com/woocommerce/woocommerce/issues/44245

**Comment by kmanijak** [comment]
> As a long-term solution, we should analyze the current status of the block, consider all areas for improvement, and ensure that the Post Excerpt has the same requirements. If yes, we can contribute upstream; otherwise, it could make sense to create a dedicated block.

I want to highlight this issue https://github.com/woocommerce/woocommerce/issues/44245.

One additional thing to consider is the WYSWIG in Product Editor. You can style the description (bold, italic, lists etc.) while it all gets flattened to the plain text in the frontend (without even a break line). It's important to mention the Classic Single Product Template supported the styling.

Having said that, I think we may need to consider creating a dedicated block because `post-excerpt` has a bit different expectations. There's this issue in Gutenberg: https://github.com/WordPress/gutenberg/issues/55562 about supporting a list but I guess it's less expected to support various styling than in Product Summary.

Product Editor | Product Summary | Single Product Classic Template
-- | -- | --
<img width="572" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/10ac7943-b160-4067-8d05-ae25d2f153fe"> | <img width="818" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/da23c304-7fd9-493a-9423-09b618b14d1f"> | <img width="546" alt="image" src="https://github.com/woocommerce/woocommerce/assets/20098064/dff7c22b-3f30-413c-b9bf-a49397064db6">

**Comment by kmanijak** [comment]
Recently I took some actions based on this discussion.

### Short-term actions

@gigitux wrote:

> As a short-term solution, I would suggest increasing the length passing a custom excerptLenght attributes ([source code](https://github.com/WordPress/gutenberg/blob/23bdd6679d73bca1f73c614644057c0f4e0c133f/packages/block-library/src/post-excerpt/block.json/#L20-L23)) when we register the Product Summary variation. ([source code](https://github.com/woocommerce/woocommerce/blob/78e38abef44f94c7a93bcafb47589bea82e5c34a/plugins/woocommerce-blocks/assets/js/blocks/product-query/variations/elements/product-summary.tsx/#L19-L26)).

After analyzing the scenarios I took a bit different approach as the block can be used in product list block (Products (Beta) and Product Collection). In those scenarios block is expected to show shorter excerpt.

Hence https://github.com/woocommerce/woocommerce/pull/47651 increases the default excerpt length to 100 words (maximum allowed by Excerpt block) in three places where Product Summary is used:
- Single Product block
- default blockified Single Product template
- the transformation of a classic Single Product template into blockified Single Product template.

I kept the default value of 55 words for Product Summary in Products (Beta) and Product Collection blocks.

In addition, I made an upstream contribution https://github.com/WordPress/gutenberg/pull/61888 fixing the problem that if a post (or product) has no excerpt, the post content is taken as a fallback. Post Content is truncated internally before reaching the Excerpt block to 55 words meaning that `excerptLenght` higher than 55 doesn't work.

### Long-term actions

@gigitux wrote:

> As a long-term solution, we should analyze the current status of the block, consider all areas for improvement, and ensure that the Post Excerpt has the same requirements. If yes, we can contribute upstream; otherwise, it could make sense to create a dedicated block.

I didn't take specific actions but did some analysis:

#### Blocks

We have four blocks/variations available:
- Excerpt ([`core/post-excerpt`](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-library/src/post-excerpt)).
- Product Summary ([`woocommerce/product-query/product-summary`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/blocks/product-query/variations/elements/product-summary.tsx) as a Products (Beta) `core/post-excerpt` variation):
    - Hidden from inserter,
    - Inherits features (and problems) of Excerpt block.
    - Replaced by below,
- Product Summary ([`woocommerce/product-collection/product-summary`](https://github.com/woocommerce/woocommerce/blob/trunk/plugins/woocommerce-blocks/assets/js/blocks/product-collection/variations/elements/product-summary.tsx) as a Product Collection `core/post-excerpt` variation):
    - Available in inserter,
    - Inherits features (and problems) of Excerpt block.
- Product Summary from  Product Elements ([`woocommerce/product-summary`].(https://github.com/woocommerce/woocommerce/tree/trunk/plugins/woocommerce-blocks/assets/js/atomic/blocks/product-elements/summary)):
    - Available for All Products and Single Product blocks
    - Missing features available in blocks above:
        - "Read more" link
        - Setting the excerpt length
    - Features available in this one and missing in blocks above
        - WYSWIG styling support

#### Recommended next steps

Considering that Excerpt has different expectations and requirements my recommendation is as follows:

1. Port missing features to the Product Summary from Product Elements:
    - "Read more" link
    - Setting the excerpt length
    - Extra (not available in Excerpt): option to disable truncation completely
2. Set Product Summary from Product Elements as a default Product Summary block everywhere (including product list blocks).

#### Misc

When writing this I realized we present two Product Summary blocks in a Single Product block. Here's the fix: https://github.com/woocommerce/woocommerce/pull/48253

**Reply by gigitux** [reply]
Thanks, @kmanijak, for working on this! Great investigation!

The `Product Summary` block is a client-side rendered block, so before adding the new features, we should implement the server-side rendering (maybe it could make sense to deprecate this one, and work a new block from scratch?)

Overall, I agree with you that we should work on a dedicated product summary block if we are not able to have more control over the variation. Last time that we discussed (p1715344646987719-slack-C02A41GCS) to continue to use the `core/post-excerpt` variation approach, we found as the main blockers:

- impossible to hide UI controls https://github.com/WordPress/gutenberg/pull/60779.
- dedicated global styles settings for block variation https://github.com/WordPress/gutenberg/pull/61481#issuecomment-2100853510.

cc @jarekmorawski in case you want to add something more!

**Reply by tjcafferkey** [reply]
>impossible to hide UI controls https://github.com/WordPress/gutenberg/pull/60779.

I think our usage of variations when it came to the `Products (Beta)` block established that using the variation API to extend functionality (such as hiding UI controls) wasn't using it for its intended purpose.

>dedicated global styles settings for block variation https://github.com/WordPress/gutenberg/pull/61481#issuecomment-2100853510.

Though as part this issue https://github.com/WordPress/gutenberg/issues/43743 global styles for variations should be addressed.

**Reply by jarekmorawski** [reply]
Thanks for the ping, folks! I think we're close to making the decision. It's better to build a dedicated block because it'll give us more flexibility in the future. As Karol said, the Excerpt in core serves a slightly different purpose and we may need to deviate from it soon.

That said, the new block should:

- offer the same functionality as the Excerpt block and more
- come with similar extensibility points
- be the only Product Summary block _ever_
- be discoverable and fully compatible with different contexts: Product Page, Single Product Block, Product Collection, Product Recommendations
- replace all other Product Summary blocks, especially those in Product Elements

**Reply by jarekmorawski** [reply]
As for adding extra settings to core blocks, I asked a relevant question in Gutenberg's internal channel: p1715344646987719-slack-C02A41GCS.

**Reply by maxlaf** [reply]
Glad to see we're moving forward with this! I agree building a dedicated block is the better solution.

Just to recap, from a Happiness perspective, these 2 things would satisfy or resolve all the cases I've seen about this in support.

>Extra (not available in Excerpt): option to disable truncation completely
>WYSWIG styling support

**Reply by kmanijak** [reply]
> The Product Summary block is a client-side rendered block, so before adding the new features, we should implement the server-side rendering (maybe it could make sense to deprecate this one, and work a new block from scratch?)

I'm against introducing yet another block while the current Product Summary gives us a good foundation. And providing SSR (which we'll have to do for block built from scratch anyway) is aligned with what happens for all of the other product elements. I'll give it a go for a day and if I find some problems/blockers I'll pivot to building a block from scratch.

But to sum up the requirements, the block:
- allow disabling truncation completely,
- allow disabling fallback to post content if the excerpt is not provided,
- support WYSIWYG from product editor,
- provide "Read more" link the same as Excerpt.

**Comment by kmanijak** [comment]
Hi folks, I consider this discussion being concluded with this PR https://github.com/woocommerce/woocommerce/pull/48984 has that will land in WooCommerce 9.6. 

**Changes include:**
- Providing improvements to the existing Product Summary block (from Product Elements):
    - support for text styling from Product Editor,
    - ability to turn on/off truncation (0 - 200 words, where 0 means no truncation),
    - ability to disable fallback to full product description,
    - "Read more" link,
    - making it compatible and working in different contexts: All Products, Single Product block, Single Product template, Product Collection.
    - adjusting compatibility layer so it takes new block into account
    - add support for languages with word definition as `characters_excluding_spaces` and `characters_including_spaces` (only on the frontend, Editor has the right implementation but the setting doesn't appear in wcBlocksSettings. It requires additional investigation: https://github.com/woocommerce/woocommerce/issues/53146)

**How to use a new block:**
- Block will be available in the inserter under "Product Summary" name
- If you're adding new blocks, the new Product Summary will be already in the markup of:
    - Single Product block
    - when transforming Classic Single Product Template into blockified version
- We don't automatically upgrade the current block (Post Excerpt's variation). So your blocks in use will work as-is but we're providing an upgrade path from current variation(s) to the new Product Summary block through Inspector Controls:

![Image](https://github.com/user-attachments/assets/239888d5-d988-4baa-985f-2d30367d24aa)

- The current Product Summary is NOT replaced in blockified Single Product template. This is a conscious decision since that would affect existing stores with a pristine Single Product template. Product Summary would look different out of nowhere and could be perceived as breaking change. If you'd like to use the new block here, replace it manually or use the upgrade path presented above.

---

The PR is merged into `trunk` so the block is ready for testing in current `trunk` or will be available in upcoming Nightly build! Any feedback is welcome! 🙌
